Clarify Auth Service backend permissions

ptgott opened the pull request:
Closes #22482

This change clarifies descriptions of:

- IAM permissions that the Auth Service requires in order to manage S3 buckets and DynamoDB tables
- When the Auth Service needs to create an S3 bucket/DynamoDB table

Specific edits:

- Add tabs to the partials we use to describe S3 and DynamoDB IAM policies. One tab item is for cases where the Auth Service creates a bucket/table, and another tab item is for when you have a pre-existing bucket/table.
- Remove an extraneous H3 in the Backends Reference re: the required DynamoDB policy. This H3 includes a very permissive policy, while another H3 in the same page includes a policy with more reasonable permissions. The more permissive H3 must have been left in the page by mistake.
- Clarify permissions in the HA guide.

**Note to reviewers:** I haven't tested this change, but made it based on studying the code the Auth Service uses to create DynamoDB tables and S3 buckets (`dynamo.New`, `s3sessions.NewHandler`). I'd appreciate your input on the accuracy of the IAM policies and table/bucket requirements in this PR.

ptgott committed:
Clarify Auth Service backend permissions

Closes #22482

This change clarifies descriptions of:

- IAM permissions that the Auth Service requires in order to manage S3
  buckets and DynamoDB tables
- When the Auth Service needs to create an S3 bucket/DynamoDB table

Specific edits:

- Add tabs to the partials we use to describe S3 and DynamoDB IAM
  policies. One tab item is for cases where the Auth Service creates a
  bucket/table, and another tab item is for when you have a pre-existing
  bucket/table.
- Remove an extraneous H3 in the Backends Reference re: the required
  DynamoDB policy. This H3 includes a very permissive policy, while
  another H3 in the same page includes a policy with more reasonable
  permissions. The more permissive H3 must have been left in the page by
  mistake.
- Clarify permissions in the HA guide.

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -126,24 +126,28 @@ events:
-or Firestore collections itself, and does not require them to exist in advance.
+For Amazon DynamoDB and Google Cloud Firestore, your Teleport configuration
+(which we will describe in more detail in the [Configuration](#configuration)
+section) names a table or collection where Teleport stores cluster state and
+audit events. 
```
It might be worth clarifying here that the tables for cluster state and audit events are always separate, as they have different schemas. I know we do mention this specifically in the partial, but I'm a fan of overstating things like this to make sure people are aware.

webvictim commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -158,20 +162,22 @@ storage services:
 objects. 
 
+Unless you plan to configure the Auth Service to access a pre-existing bucket,
+you must also assign Auth Service instances permissions to create buckets
```
```suggestion
you must also assign Auth Service instances permissions to create buckets.
```

webvictim commented on docs/pages/includes/dynamodb-iam-policy.mdx:
```diff
@@ -1,3 +1,99 @@
+<Tabs>
+<TabItem label="Manage a Table Yourself">
+
+If you choose to manage a DynamoDB table yourself, the table must have the
+following attribute definitions:
+
+|Name|Type|
+|---|---|
+|`HashKey`|`S`|
+|`FullPath`|`S`|
+
+The table must also have the following key schema elements:
+
+|Name|Type|
+|---|---|
+|`HashKey`|`HASH`|
+|`FullPath`|`RANGE`|
```
I'm quite wary of putting this specific information in the guide as they it may not be updated if we do a schema migration in future. We got burnt badly by this in the past with Terraform (https://github.com/gravitational/teleport/issues/7306)

I wonder whether maybe we could link to the code which generates the tables instead so people can read for themselves if they insist on having full control?

ptgott replied:
I can see why we would need to prevent the impression that these schema requirements are permanent, but for me, docs should prevent users from needing to look up the source. I'm not sure, for example, if the source for generating the tables will always be in a single place or easy to follow. 

What would you think of advising users to set up the backend (and upgrade their clusters) in a non-prod environment first to ensure that schema changes don't impact users?

ptgott replied:
@webvictim Just following up on this thread. Thanks!

ptgott replied:
👍 This change is meant to accommodate admins who are worried about infrastructure drift (https://github.com/gravitational/teleport/issues/19974) by having Teleport create tables. I'll add a comment in the source

ptgott committed:
Respond to webvictim feedback

alexfornuto reviewed (approved):
Some notes and nits, otherwise the copy looks good.

ptgott commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -126,24 +126,28 @@ events:
+
 <Admonition title="Required permissions">
 
 In your cloud provider's RBAC solution (e.g., AWS or Google Cloud IAM), your
 Auth Service instances need permissions to read from and write to your chosen
```
Rephrased this a bit

ptgott commented on docs/pages/deploy-a-cluster/high-availability.mdx:
```diff
@@ -126,24 +126,28 @@ events:
+key/value store. 
+
+Unless you have a pre-existing table or collection for the Auth Service to use,
+it must also have permissions to create tables and collections (if your
```
Line 149 renders as being within the same paragraph as line 148, so I think "it" works here. Otherwise, the sentence says, "Unless you have a pre-existing table or collection for the Auth Service to use, the Auth Service must also have permissions to create tables and collections"

alexfornuto replied:
Ahh, I thought "it" might refer to another noun from the previous paragraph.

ptgott replied:
Looking at this again now, I can definitely see how "it" could be clearer. I've suggested another wording.

ptgott commented on docs/pages/includes/dynamodb-iam-policy.mdx:
```diff
@@ -1,3 +1,99 @@
+Service to create and configure one for you:
+
+<Tabs>
+<TabItem label="Manage a Table Yourself">
```
I don't think we have a standard for this. On this particular page, all TabItem labels are title-cased.

ptgott committed:
Respond to alexfornuto feedback

ptgott committed:
Respond to webvictim feedback

Add a comment above the function that creates a DynamoDB table to change
the docs if the schema changes.

ptgott committed:
Respond to alexfornuto feedback
